File: step_12_greedy/12.2_medium_hard/P011_non_overlapping_intervals/solution.py
"""
Problem: Non Overlapping Intervals
LeetCode 435 | Difficulty: MEDIUM | XP: 25

Given an array of intervals, return the minimum number of intervals you need
to remove to make the rest non-overlapping.

Key Insight: Equivalent to finding the MAXIMUM number of non-overlapping
             intervals (like Activity Selection), then answer = n - max_kept.
             Greedy: sort by end time, always keep the interval that ends earliest.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE  (Try all subsets)
# Time: O(2^n * n)  |  Space: O(n)
# ============================================================
def brute_force(intervals: List[List[int]]) -> int:
    """
    Generate every subset of intervals and check if it is non-overlapping.
    Track the maximum non-overlapping subset size.
    Answer = n - max_non_overlapping_size.
    Only feasible for very small n.
    """
    from itertools import combinations

    n = len(intervals)

    def is_non_overlapping(subset):
        sorted_sub = sorted(subset, key=lambda x: x[0])
        for i in range(1, len(sorted_sub)):
            if sorted_sub[i][0] < sorted_sub[i - 1][1]:
                return False
        return True

    max_kept = 0
    for size in range(n, -1, -1):
        for combo in combinations(intervals, size):
            if is_non_overlapping(list(combo)):
                max_kept = max(max_kept, size)
                break
    return n - max_kept


# ============================================================
# APPROACH 2: OPTIMAL  (Greedy — sort by end time)
# Time: O(n log n)  |  Space: O(1)
# ============================================================
def optimal(intervals: List[List[int]]) -> int:
    """
    Sort intervals by their END time.
    Greedily keep each interval whose start >= last kept end.
    This maximises the number kept (Activity Selection Theorem).
    Answer = total intervals - number kept.
    """
    if not intervals:
        return 0
    intervals.sort(key=lambda x: x[1])
    kept = 0
    last_end = float('-inf')
    for start, end in intervals:
        if start >= last_end:
            kept += 1
            last_end = end
    return len(intervals) - kept


# ============================================================
# APPROACH 3: BEST  (same greedy, count removals directly)
# Time: O(n log n)  |  Space: O(1)
# ============================================================
def best(intervals: List[List[int]]) -> int:
    """
    Equivalent greedy, but count removals directly instead of kept intervals.
    When an overlap is detected (current start < last_end), increment removals
    and keep the interval with the smaller end (to leave more room for future).
    """
    if not intervals:
        return 0
    intervals.sort(key=lambda x: x[1])
    removals = 0
    last_end = float('-inf')
    for start, end in intervals:
        if start < last_end:
            removals += 1
            # Keep the one with smaller end (already sorted so current end >= last_end;
            # last_end is already the smaller one — no update needed)
        else:
            last_end = end
    return removals


if __name__ == "__main__":
    test_cases = [
        ([[1, 2], [2, 3], [3, 4], [1, 3]], 1),
        ([[1, 2], [1, 2], [1, 2]], 2),
        ([[1, 2], [2, 3]], 0),
        ([], 0),
        ([[1, 100], [11, 22], [1, 11], [2, 12]], 2),
    ]
    print("=== Non Overlapping Intervals ===")
    for intervals, expected in test_cases:
        b   = brute_force([iv[:] for iv in intervals])
        o   = optimal([iv[:] for iv in intervals])
        bst = best([iv[:] for iv in intervals])
        status = "OK" if b == o == bst == expected else "FAIL"
        print(f"  intervals={intervals} => brute={b}, optimal={o}, best={bst} "
              f"(expected {expected}) [{status}]")

File: step_12_greedy/12.2_medium_hard/P011_non_overlapping_intervals/Solution.java
/**
 * Problem: Non Overlapping Intervals
 * LeetCode 435 | Difficulty: MEDIUM | XP: 25
 *
 * Given an array of intervals, return the minimum number of intervals to remove
 * to make the remaining intervals non-overlapping.
 *
 * Key Insight: Equivalent to finding the maximum set of non-overlapping intervals
 *              (Activity Selection), then answer = n - max_kept.
 *              Greedy: sort by end time, keep earliest-ending intervals.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE  (Try all subsets)
    // Time: O(2^n * n)  |  Space: O(n)
    // ============================================================
    static class BruteForce {
        /**
         * Generate all 2^n subsets. For each subset, check if it is
         * non-overlapping (sort by start, verify no two intervals overlap).
         * Track the maximum non-overlapping subset size.
         * Answer = n - max_kept.
         */
        public int eraseOverlapIntervals(int[][] intervals) {
            int n = intervals.length;
            int maxKept = 0;
            for (int mask = 0; mask < (1 << n); mask++) {
                List<int[]> subset = new ArrayList<>();
                for (int i = 0; i < n; i++) {
                    if ((mask >> i & 1) == 1) subset.add(intervals[i]);
                }
                if (isNonOverlapping(subset)) {
                    maxKept = Math.max(maxKept, subset.size());
                }
            }
            return n - maxKept;
        }

        private boolean isNonOverlapping(List<int[]> subset) {
            subset.sort(Comparator.comparingInt(a -> a[0]));
            for (int i = 1; i < subset.size(); i++) {
                if (subset.get(i)[0] < subset.get(i - 1)[1]) return false;
            }
            return true;
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL  (Greedy — sort by end time)
    // Time: O(n log n)  |  Space: O(1)
    // ============================================================
    static class Optimal {
        /**
         * Sort intervals by end time.
         * Greedily keep any interval whose start >= lastEnd.
         * This maximises the count of non-overlapping intervals kept
         * (Activity Selection Theorem proof: exchange argument).
         * Answer = n - kept.
         */
        public int eraseOverlapIntervals(int[][] intervals) {
            if (intervals.length == 0) return 0;
            Arrays.sort(intervals, Comparator.comparingInt(a -> a[1]));
            int kept = 0;
            int lastEnd = Integer.MIN_VALUE;
            for (int[] iv : intervals) {
                if (iv[0] >= lastEnd) {
                    kept++;
                    lastEnd = iv[1];
                }
            }
            return intervals.length - kept;
        }
    }

    // ============================================================
    // APPROACH 3: BEST  (Same greedy, count removals directly)
    // Time: O(n log n)  |  Space: O(1)
    // ============================================================
    static class Best {
        /**
         * Count removals directly: when an interval overlaps with the last
         * kept interval (current start < lastEnd), increment removals.
         * We keep the interval with the smaller end (already sorted ascending
         * by end, so lastEnd is always the smaller — no update on removal).
         */
        public int eraseOverlapIntervals(int[][] intervals) {
            if (intervals.length == 0) return 0;
            Arrays.sort(intervals, Comparator.comparingInt(a -> a[1]));
            int removals = 0;
            int lastEnd = Integer.MIN_VALUE;
            for (int[] iv : intervals) {
                if (iv[0] < lastEnd) {
                    removals++;   // remove current (it ends later, worse for future)
                } else {
                    lastEnd = iv[1];
                }
            }
            return removals;
        }
    }

    public static void main(String[] args) {
        int[][][] tests = {
            {{1,2},{2,3},{3,4},{1,3}},
            {{1,2},{1,2},{1,2}},
            {{1,2},{2,3}},
            {{1,100},{11,22},{1,11},{2,12}}
        };
        int[] exps = {1, 2, 0, 2};

        BruteForce bf  = new BruteForce();
        Optimal    opt = new Optimal();
        Best       bst = new Best();

        System.out.println("=== Non Overlapping Intervals ===");
        for (int t = 0; t < tests.length; t++) {
            int b  = bf.eraseOverlapIntervals(tests[t]);
            int o  = opt.eraseOverlapIntervals(tests[t]);
            int be = bst.eraseOverlapIntervals(tests[t]);
            String status = (b == o && o == be && be == exps[t]) ? "OK" : "FAIL";
            System.out.printf("  brute=%d opt=%d best=%d (exp=%d) [%s]%n",
                              b, o, be, exps[t], status);
        }
    }
}
